Support for nif:ContextHashBasedString

Ga11u opened the pull request:
- All tests passed
- Added 3 more tests for nif:ContextHashBasedString
- Updated README.md
- Modifications applied to:
-- context.py
-- phrase.py
-- test_context.py
-- test_phrase.py

This new version allows passing nif:ContextHashBasedString as :param: ``uri``. I added an extra parameter ``is_hash_based_uri`` to the declaration of Context and Phrases classes. The  ``is_hash_based_uri`` parameter indicates whether the uri is ContextHashBasedString or OffsetBasedString (``True`` or ``False`` respectively). If the  ``is_hash_based_uri`` is ``True``, then the resulting rdf serialization uses ``nif:ContextHashBasedString``  instead of ``nif:OffsetBasedString``.

The function ``def turtle(self)`` was also updated on both NIFContext and NIFPhrase. I added as it is in the NIFCollection:
```py
        # workaround for https://github.com/RDFLib/rdflib/issues/884
        if isinstance(out, binary_type):
            out = out.decode('utf-8')
        return out
```

**Unresolved**
The nif:ContextHashBasedString URI has its own format described in Linked-Data Aware URI Schemes for Referencing Text Fragments (https://doi.org/10.1007/978-3-642-33876-2_17) page 4. The user must provide a URI formated as ContextHashBasedString, the pynif does not generate URI formated as ContextHashBasedString.

- A posible enhancment for pynif library would be adding a module for creating ContextHashBasedString formated URI.


**Solution to a know possible issue**
The default behaviour of pyinf creates ``OffsetBasedString`` for ``Phrases`` if a ``uri`` is **not** passed as a parameter. **This behaviour has not been changed.** Hence, if the ``Context`` is using ``ContextHashBasedString`` and a ``uri`` is not passed as parameter when a ``phrase`` is created using the ``Context.add_phrase( ... )`` or ``NIFPhrase( ... )`` function , the resulting phrase' URI will have an ``OffsetBasedString`` instead of a ``ContextHashBasedString``.  The resulting ``OffsetBasedString`` will be based on the ``:param: context`` value + #offset_X_Y which can end up creating a URI with mixed parts from the ``ContextHashBasedString`` and #offset_X_Y. 

- Possible solution:

1. **Current solution**: The user is responsible for passing the correct parameters. If done correctly, there is not any odd behaviour.
2. Define some assertions based on the value ``is_hash_based_uri``. For example, if ``is_hash_based_uri`` is ``True``, then the function  ``Context.add_phrase( ... )`` must have a ``uri`` value != None.

Ga11u committed:
modified:   .gitignore
	modified:   pynif/context.py
	modified:   pynif/phrase.py
	modified:   pynif/tests/test_collection.py
	modified:   pynif/tests/test_context.py
	modified:   pynif/tests/test_phrase.py

Ga11u committed:
modified:   README.md
	modified:   pynif/tests/test_context.py

Ga11u committed:
deleted:    .vscode/settings.json
	modified:   README.md

Ga11u committed:
modified:   README.md

Ga11u committed:
modified:   pynif/tests/test_phrase.py

Ga11u committed:
modified:   pynif/context.py
	modified:   pynif/phrase.py
	modified:   pynif/tests/test_context.py
	modified:   pynif/tests/test_phrase.py

Ga11u committed:
modified:   README.md

Ga11u committed:
modified:   pynif/context.py
	modified:   pynif/phrase.py

Ga11u committed:
modified:   README.md

wetneb reviewed (commented):
Very nice! I can think of another use case to test: if I parse a NIF file which contains some ContextHashBasedStrings, then they should correctly be detected as such (`isContextHashBasedString`) should be true on them.
Another way to test this would be to do round-trip parsing (first parsing and then rendering in RDF again, and checking that the result is identical to the input) with some ContextHashBasedStrings in them.

Ga11u committed:
modified:   pynif/context.py
	modified:   pynif/phrase.py
	new file:   pynif/tests/data/example-contextHashBasedString.ttl
	modified:   pynif/tests/test_context.py

Ga11u committed:
modified:   pynif/tests/test_context.py

Ga11u committed:
modified:   pynif/tests/test_context.py

Ga11u committed:
modified:   pynif/tests/test_context.py

Ga11u committed:
modified:   pynif/tests/test_context.py

Ga11u committed:
modified:   pynif/tests/test_context.py

wetneb commented on pynif/tests/test_context.py:
```diff
@@ -20,7 +22,51 @@ def setUpClass(cls):
+        # The following string is copied from the file tests/data/example-ContextHashBasedString.ttl
+        # The load from file would not pass GitHub Checks sinse the file example-ContextHashBasedString.ttl 
+        # it is not in the main repo. Once the PR is completed the  line 
+        # cls.example_ContextHashBasedString = """....""" can be replaced by the following code:
```
I might have misunderstood something, but why cannot you commit the file to your branch? This should make it available in the tests and would add it to the main repo once we merge the PR.

Ga11u replied:
For some reason the GitHub Check gives that error:
test_collection.py:
> testdir = os.path.dirname(os.path.abspath(__file__))
> with open(os.path.join(testdir, 'data/example-ContextHashBasedString.ttl'), 'r') as f:
>       cls.example_ContextHashBasedString = f.read()
>
> E FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/pynif/pynif/pynif/tests/data/example-ContextHashBasedString.ttl'

The file is already commit and pushed to my own repo. That is why I think the error comes because the file is not in  wetneb/pynif repo.

Ga11u replied:
This is the error:
        testdir = os.path.dirname(os.path.abspath(__file__))
>       with open(os.path.join(testdir, 'data/example-ContextHashBasedString.ttl'), 'r') as f:
E       FileNotFoundError: [Errno 2] No such file or directory: '/home/runner/work/pynif/pynif/pynif/tests/data/example-ContextHashBasedString.ttl'

wetneb replied:
I think you just haven't added it to your commits. If you are using the command-line, `git add tests/data/example-ContextHashBasedString.ttl`, `git commit` and `git push` should do. If you are using some GUI, then there should be an option to add untracked files.

Ga11u replied:
Solved! It worked now. It was a case sensitive problem. ``data/example-ContextHashBasedString.ttl`` has a capital C and it is not. I just changed to ``data/example-contextHashBasedString.ttl`` and it work

Ga11u commented:
> 
> 
> Very nice! I can think of another use case to test: if I parse a NIF file which contains some ContextHashBasedStrings, then they should correctly be detected as such (`isContextHashBasedString`) should be true on them.
> Another way to test this would be to do round-trip parsing (first parsing and then rendering in RDF again, and checking that the result is identical to the input) with some ContextHashBasedStrings in them.

@wetneb Goid point! I added those tests too: (1) manually creating a NIFContext with phrases and comparing its turtle represention with a NIF file,  (2) manually creating a NIFContext with phrases and comparing its turtle represention with a loaded graph from NIF file, (3) loading a graph from a NIF file, trasnforming it and checking that isContextHashBasedString is present in both context and all phrases.

Ga11u committed:
modified:   pynif/tests/test_context.py

Ga11u committed:
modified:   pynif/tests/test_context.py

Ga11u committed:
modified:   pynif/tests/test_context.py

Ga11u commented:
@wetneb what other tests do you think could be interesting to implement?

wetneb commented on pynif/tests/test_context.py:
```diff
@@ -20,7 +22,48 @@ def setUpClass(cls):
+    nif:isString "Diego Maradona is from Argentina." .
+
+"
+"""
```
It looks like this big string is no longer needed and can be removed.

Ga11u replied:
Yes, you are right! I also went through the code again to check for any other string or unsed text. The only remaining text are the nif:ContextHashBasedString explanations at context.py and phrase.py .
Also added some more assertions on the test and normalized the style according to the previous tests.

Ga11u committed:
modified:   test_context.py

Ga11u committed:
modified:   test_context.py
	modified:   test_phrase.py
